Handle UNC paths in windows (fixes #23)

gajus opened the pull request.

gajus commented:
@sokra Please review this, as it affects the logic outside of just pathToArray (normalisation).

sokra commented on lib/MemoryFileSystem.js:
```diff
@@ -38,8 +38,16 @@ function isFile(item) {
 function pathToArray(path) {
 	path = normalize(path);
+
+	var UNC = /^\\\\/.test(path);
```
lower case `unc`. all upper case only for constants.

sokra commented:
needs test cases for normalization

i. e. `fs.normalize("\\\\a\\\\b/../c\\d\\..").should.be.eql("\\\\a\\c");`

gajus commented:
@sokra are you certain regarding addition/support of this test case?

```js
fs.normalize("\\\\a\\\\b/../c\\d\\..").should.be.eql("\\\\a\\c");
```

It would make the code quite ugly. That is, I had a few dabs at it, and reverted. The fact that in *nix a file/ directory name can contain a backslash (such as in the case of [this test](https://github.com/webpack/memory-fs/blob/d248b04aa7b48146d98a110df182131bab531793/test/MemoryFileSystem.js#L23)), makes it particularly ugly.

Besides, none of the existing test cases would support an equivalent mix of *nix and windows DS, e.g. these will fail:

```js
"/a/b/c/..\\.."
// AssertionError: expected '/a/b/c/..\\..' to equal '/a'

"C:\\a\\b\\\c\\../.."
// AssertionError: expected 'C:\\a\\b\\c\\../..' to equal 'C:\\a'
```

am I overlooking something?

sokra commented:
oh yeah, I was wrong. the normalization does only care for `.` and `..`. The path joining corrects the slashes. Here the [`absoluteWinRegexp`](https://github.com/webpack/memory-fs/blob/master/lib/join.js#L3) need to include UNC paths.

gajus commented:
> Here the absoluteWinRegexp need to include UNC paths.

Whats the need for the addition?

What test case am I missing?

https://github.com/webpack/memory-fs/pull/28/commits/407053104022c5cf90b640efe04bdd21073d8349

As it is, all scenarios are covered, as far as my testing goes.

gajus commented:
@sokra I had a second look into this. I don't see what you want to achieve by adding the UNC path support to `absoluteWinRegexp`. UNC paths are already handled as expected, except in cases when there is a mix of nix and windows DS, which does not work with the current test suite anyway.

I am happy to add the support for mixing windows/nix DS, though as I have mentioned earlier, it will require some refactoring and it is not going to work in all cases, since nix paths can contain backwards slashes in file/directory names.

sokra commented:
mixed windows/nix paths will occur. i. e. when resolving `./relative/path` in `\\srv\share\src`.
But windows style path separators will not appear in linux paths.

So `\\srv\share\src/relative/path` is possible, but `/root/src\relative\path` is not possible (it's possible but `\` is not a path separator in this case). Because of this lucky circumstance we do not have a problem with backwards slashes in linux paths.

sokra commented on lib/normalize.js:
```diff
@@ -34,5 +35,13 @@ module.exports = function normalize(path) {
+
+	if (unc) {
+		path = '\\' + path;
+	}
```
instead of this block, could we just do:

``` js
var doubleSlashWinRegExp = /(?!^)\\+/g;
```

gajus replied:
a66aa97092e87cfa838126bc66b94a39ffeeba0b

gajus commented:
> oh yeah, I was wrong. the normalization does only care for . and ... The path joining corrects the slashes. Here the absoluteWinRegexp need to include UNC paths.

3ccfa79

gajus commented:
> So \\srv\share\src/relative/path is possible, but /root/src\relative\path is not possible (it's possible but \ is not a path separator in this case). Because of this lucky circumstance we do not have a problem with backwards slashes in linux paths.

I have fixed the refactoring issue and fixed the join function for use with UNC paths. However, I am leaving the normalisation of a mix of path to issue open. I don't see how it can be solved without major refactoring of normalise.js.
